network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -537,90 +543,250 @@ def pretty_print_diff(diff_list):
+        return result.replace('{', ' ').replace('}', ' ').replace('\'', '')
 
-        print("Executing semantic diff")
-        peers_to_compare = self.config1.peer_container.get_all_peers_group()
-        peers_to_compare |= self.disjoint_ip_blocks()
-        captured_pods = self.config1.get_captured_pods() | self.config2.get_captured_pods()
-        all_diffs = []
-        for pod1 in peers_to_compare:
-            for pod2 in peers_to_compare if pod1 in captured_pods else captured_pods:
+    def compute_diff(self):
+        """
+        Compute changed connections as following:
+
+        1.1. lost connections between removed peers
+        1.2. lost connections between removed peers and old ipBlock peers
+
+        2.1. lost connections between removed peers and intersected peers
+
+        3.1. lost/new connections between intersected peers due to changes in policies and labels of pods/namespaces
+        3.2  lost connections between intersected peers and removed ipBlock peers
+        3.3  lost/new connections between intersected peers and intersected pBlock peers due to changes in policies and labels of pods/namespaces
+        3.4  new connections between intersected peers and added ipBlock peers
+
+        4.1. new connections between intersected peers and added peers
+
+        5.1. new connections between added peers
+        5.2. new connections between added peers and new ipBlock peers
+
+        Some of the section might be empty and can be dropped.
+
+        :return: A dictionary with lists of all various diffs
+        :rtype: dict
+        """
+        all_diff = {}
+        old_peers = self.config1.peer_container.get_all_peers_group()
+        new_peers = self.config2.peer_container.get_all_peers_group()
+        intersected_peers = old_peers & new_peers
+        removed_peers = old_peers - intersected_peers
+        added_peers = new_peers - intersected_peers
+
+        captured_pods = (self.config1.get_captured_pods() | self.config2.get_captured_pods()) & intersected_peers
+
+        # compute 3 disjoint sets that cover all referenced ipBlock-peers in old
+        # and new configurations: removed-peers, intersected-peers, added-peers
+        disjoint_ip_blocks = self.disjoint_ip_blocks()
+
+        # 1.1. lost connections between removed peers
+        key = 'Lost connections between removed peers'
+        all_diff[key] = []
+        for pair in itertools.permutations(removed_peers, 2):
+            _, lost_conns, _ = self.config1.allowed_connections(pair[0], pair[1])
+            if str(lost_conns) != 'No connections':
+                all_diff[key].append(
+                    SemanticDiffQuery.SingleDiff(pair[0], pair[1], lost_conns, None))
+
+        # 1.2. lost connections between removed peers and old ipBlock peers
+        key = 'Lost connections between removed peers and old ipBlock peers'
+        all_diff[key] = []
+        for pair in itertools.product(removed_peers, disjoint_ip_blocks):
```

Comment:
```suggestion
        for pair in itertools.product(removed_peers, self.config1.get_referenced_ip_blocks()):
```

Reply:
No, because in case no ipBlocks are specified then `self.config1.get_referenced_ip_blocks()` returns empty set, yet for any uncovered pods traffic to/from ipBlocks is allowed.

Reply:
You are correct. I forgot about this case.
However, better make the case where `self.config1.get_referenced_ip_blocks()` returns an empty set a special case, and use the whole ip range instead.
e.g.,:
```
config1_ips = self.config1.get_referenced_ip_blocks() or IpBlock.get_all_ips_block()
```

Reply:
this is still not exact.
take the case where `self.config1.get_referenced_ip_blocks()` are specified. then for removed captured pods this would work as required, however for removed uncaptured pods we will be missing connectivity with ipBlocks that are not within `self.config1.get_referenced_ip_blocks()`.
If we want to take this direction then we'll need to distinguish between:
1. for removed captured-pods check against `self.config1.get_referenced_ip_blocks()` 
2. for removed uncaptured-pods check against `IpBlock.get_all_ips_block()`

or keep existing implementation using `disjoint_ip_blocks`.
I think this one is more clean programming, but would change with any option you think best.

Reply:
You are right. What we really need is using the result of applying `disjoint_ip_blocks()` to `self.config1.get_referenced_ip_blocks()` and to `IpBlock.get_all_ips_block()`.
How about making `disjoint_ip_blocks()` parametric so we can accomplish this?

Reply:
good catch - fixed. thanks.